pulsar/producer_partition.go

```diff
@@ -441,30 +446,33 @@ func (p *partitionProducer) reconnectToBroker() {
 		}
 		p.log.WithError(err).Error("Failed to create producer at reconnect")
 		errMsg := err.Error()
-		if strings.Contains(errMsg, errTopicNotFount) {
+		if strings.Contains(errMsg, errMsgTopicNotFound) {
 			// when topic is deleted, we should give up reconnection.
-			p.log.Warn("Topic Not Found.")
+			p.log.Warn("Topic not found, stop reconnecting")
 			break
 		}
 
-		if strings.Contains(errMsg, "TopicTerminatedError") {
-			p.log.Info("Topic was terminated, failing pending messages, will not reconnect")
-			pendingItems := p.pendingQueue.ReadableSlice()
-			for _, item := range pendingItems {
-				pi := item.(*pendingItem)
-				if pi != nil {
-					pi.Lock()
-					requests := pi.sendRequests
-					for _, req := range requests {
-						sr := req.(*sendRequest)
-						if sr != nil {
-							sr.done(nil, newError(TopicTerminated, err.Error()))
-						}
-					}
-					pi.Unlock()
-				}
-			}
-			p.setProducerState(producerClosing)
+		if strings.Contains(errMsg, errMsgTopicTerminated) {
+			p.log.Warn("Topic was terminated, failing pending messages, stop reconnecting")
+			p.failPendingMessages(newError(TopicTerminated, err.Error()))
+			// can not set to producerClosing , or it will fail when we call internalClose()
+			// there is a Terminated state in JAVA client, maybe we should add a producerTerminated state ?
+			// p.setProducerState(producerClosing)
```

Comment:
> or it will fail when we call internalClose()

IMO, we should set the producer state to a terminated state here. The `internalClose()` should be skipped when the state is in a terminated state(TopicTerminated, Fenced, etc).

Reply:
@RobertIndie @tisonkun I have update the PR, just close the producer when TopicNotFound/TopicTerminated/ProducerFenced, would you please check if it is enough.  

I haven't imported any new state(TopicTerminated/ProducerFenced), 'cause there are too many states in JAVA client, see below, making the Go client and the JAVA client exactly like will be a BIG change, I am not sure I can do that well.

``` JAVA
 enum State {
        Uninitialized, // Not initialized
        Connecting, // Client connecting to broker
        Ready, // Handler is being used
        Closing, // Close cmd has been sent to broker
        Closed, // Broker acked the close
        Terminated, // Topic associated with this handler
                    // has been terminated
        Failed, // Handler is failed
        RegisteringSchema, // Handler is registering schema
        ProducerFenced, // The producer has been fenced by the broker
    };
```